tests/sensors/resources/generic-sensor-tests.js

```diff
@@ -44,54 +77,90 @@ function runGenericSensorTests(sensorType, verifyReading) {
-        sensor.stop();
-        t.done();
-      });
-    }, 1000);
+    sensor.onchange = t.step_func_done(() => {
+      //sensor.timestamp changes.
+      let cachedTimeStamp2 = sensor.timestamp;
+      assert_greater_than(cachedTimeStamp2, cachedTimeStamp1);
+      sensor.stop();
+      t.done();
+    });
   }, "timestamp is updated when time passes");
 
   test(() => {
-    let sensor, start_return;
+    let sensor;
     sensor = new sensorType();
-    sensor.onerror = () => {
-      assert_unreached(event.error.name + ":" + event.error.message);
-    };
-    //The default sensor.state is 'unconnected'
+    sensor.onerror = unreached;
     assert_equals(sensor.state, "unconnected");
-    start_return = sensor.start();
-    //The sensor.state changes to 'activating' after sensor.start()
+  }, "default sensor.state is 'unconnected'");
+
+  test(() => {
+    let sensor;
+    sensor = new sensorType();
+    sensor.onerror = unreached;
+    sensor.start();
     assert_equals(sensor.state, "activating");
-    //TODO: The permission is not ready.
-    //the sensor.start() return undefined
+    sensor.stop();
+  }, "sensor.state changes to 'activating' after sensor.start()");
+
+  test(() => {
+    let sensor, start_return;
+    sensor = new sensorType();
+    sensor.onerror = unreached;
+    start_return = sensor.start();
     assert_equals(start_return, undefined);
     sensor.stop();
-  }, "sensor.start() is correct");
+  }, "sensor.start() return undefined");
+
+  test(() => {
+    try {
+      let sensor;
+      sensor = new sensorType();
+      sensor.onerror = unreached;
+      sensor.start();
+      sensor.start();
+      assert_equals(sensor.state, "activating");
+      sensor.stop();
+    } catch (e) {
+       assert_unreached(e.name + ": " + e.message);
+    }
+  }, "sensor.start() methods do not throw");
+
+  test(() => {
+    let sensor;
+    sensor = new sensorType();
+    sensor.onerror = unreached;
+    sensor.start();
+    sensor.stop();
+    assert_equals(sensor.state, "idle");
+  }, "sensor.state changes to 'idle' after sensor.stop()");
 
   test(() => {
     let sensor, stop_return;
     sensor = new sensorType();
-    sensor.onerror = () => {
-      assert_unreached(event.error.name + ":" + event.error.message);
-    };
+    sensor.onerror = unreached;
     sensor.start();
     stop_return = sensor.stop();
-    //The sensor.state changes to 'idle' after sensor.stop()
-    assert_equals(sensor.state, "idle");
-    //the sensor.stop() returns undefined
     assert_equals(stop_return, undefined);
-  }, "sensor.stop() is correct");
+  }, "sensor.stop() returns undefined");
+
+  test(() => {
+    try {
+      let sensor;
+      sensor = new sensorType();
+      sensor.onerror = unreached;
+      sensor.start();
+      sensor.stop();
+      sensor.stop();
+      assert_equals(sensor.state, "idle");
+    } catch (e) {
+       assert_unreached(e.name + ": " + e.message);
+    }
+  }, "sensor.stop() methods do not throw");
```

Comment:
use "No exception threw when calling stop() on already stopped sensor" would be better